src/hypothesis/strategies.py

```diff
@@ -836,31 +836,75 @@ def dm_func(denom):
-    and NaN are possible values generated by this strategy.
+    Note that where floats have one `NaN` value, Decimals have four: signed,
+    and either *quiet* or *signalling*.  See `the decimal module docs
+    <https://docs.python.org/3/library/decimal.html#special-values>`_ for
+    more information on special values.
 
     """
-    floats_strategy = floats(
-        min_value=min_value, max_value=max_value,
-        allow_nan=allow_nan, allow_infinity=allow_infinity)
-    fractions_strategy = fractions(min_value=min_value, max_value=max_value)
-
-    return (
-        floats_strategy.map(float_to_decimal) |
-        fractions_strategy.map(lambda f: Decimal(f.numerator) / f.denominator)
-    )
+    # Convert min_value and max_value to Decimal values, and validate args
+    check_valid_integer(places)
+    if places is not None and places < 0:
+        raise InvalidArgument('cannot have negative number of places (%r)')
+    try:
+        min_value = Decimal(min_value)
+    except:
+        pass
+    try:
+        max_value = Decimal(max_value)
+    except:
+        pass
+    if isinstance(min_value, Decimal) and not min_value.is_finite():
+        raise InvalidArgument('min_value=%r is not a finite bound' % min_value)
+    if isinstance(max_value, Decimal) and not max_value.is_finite():
+        raise InvalidArgument('max_value=%r is not a finite bound' % max_value)
+    check_valid_bound(min_value, 'min_value')
+    check_valid_bound(max_value, 'max_value')
+    check_valid_interval(min_value, max_value, 'min_value', 'max_value')
+    if allow_infinity is None:
+        allow_infinity = min_value is not None or max_value is not None
+    elif allow_infinity and min_value is not None and max_value is not None:
+        raise InvalidArgument('Cannot allow infinity between finite bounds')
+    # Set up a strategy for finite decimals
+    if places is not None:
+        # Fixed-point decimals are basically integers with a scale factor
+
+        def try_quantize(d):
+            try:
+                return d.quantize(factor)
+            except InvalidOperation:  # pragma: no cover
+                return None
+        factor = Decimal(10) ** -places
+        max_num = max_value / factor if max_value is not None else None
+        min_num = min_value / factor if min_value is not None else None
+        strat = integers(min_value=min_num, max_value=max_num)\
+            .map(lambda d: try_quantize(d * factor))\
+            .filter(lambda d: d is not None)
+    else:
+        # Otherwise, they're like fractions featuring a power of ten
+        strat = fractions(
+            min_value=min_value, max_value=max_value
+        ).map(lambda f: Decimal(f.numerator) / f.denominator)
+    # Compose with sampled_from for infinities and NaNs as appropriate
+    special = []
+    if allow_nan:
+        special.extend(map(Decimal, ('NaN', '-NaN', 'sNaN', '-sNaN')))
+    if allow_infinity:
+        if max_value is None:
```

Comment:
(The right way to handle this might be to convert such values to None up at the top sot hat it also works correctly when passing bounds to other strategies)